### PR TITLE
Re #19951:  ISIS Reflectometry: Prevent external table modification while processing

### LINKED ISSUE #19951: ISIS Reflectometry: Prevent external table modification while processing
While the current interface greys out the `process` buttons and prevents window closure while processing runs, modification to the table of data is not prevented and can cause errors when done. These include:

- Adding/deleting rows/groups
- Grouping rows
- Cutting/pasting rows
- Editing row data
- Transferring rows

In addition, the whole `Reflectometry` menu (except for `Export .TBL` and `Slit Calculator`) should be disabled (#21215). Furthermore, one should not be able to select any run in the table while processing occurs.

Edit: Another bug I want to address in this is that currently, pausing reduction will immediately make the `process` button immediately available (before reduction itself has actually paused). If `process` is clicked before reduction has finished, a critical error would be thrown. `Process` should be re-enabled only when reduction has finished, not just when the pause button is clicked. This has been noted as part of #21167.

In summary we need to:
- [x] Disable editing rows in the data processor widget
- [x] Disable relevant items in the menus #21215

### PR DESCRIPTION
**Pull request by edwardb96**
- Disables actions which modify the contents of the DataProcessorWidget runs table whilst the widget is processing and not paused.
- Refactors elements of the `QDataProcessorWidget`, `QtReflRunsTabView`, `ReflRunsTabPresenter` and `GenericDataProcessor` as well as the tests `ReflRunsTabPresenterTest`.

**To test:**
- Unit Tests:
  - Compile and run the test targets `MantidWidgetsTest` and `CustomInterfacesTest`
- Manual Tests:
  - Login into the ICAT catalog.
  - The interface can be accessed from Interfaces->Reflectometry->ISIS Reflectometry.
  - Select `POLREF`, enter investigation id `1520249` and click on the `autoreduce` button to begin autoreduction.
  - Ensure that it is not possible to perform actions which edit the contents of table while processing through either the Data Processor Widget toolbar, the context (right click) menu, or the 'Edit' and 'Reflectometry' menus.
  - Ensure that the actions disabled while processing are made available again when the processing is paused and when the processing is completed.

Fixes #19951. 

**Release Notes** 
*Does not need to be in the release notes?*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/ISISReflectometry/ReflectometryAction.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+File change history is stored at: <https://github.com/mantidproject/mantid>.
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+enum class ReflectometryAction {
+  OPEN_TABLE,
+  NEW_TABLE,
+  SAVE_TABLE,
+  SAVE_TABLE_AS,
+  IMPORT_TBL,
+  EXPORT_TBL
```
Comment (AntonPiccardoSelg) [inline qt/scientific_interfaces/ISISReflectometry/ReflectometryAction.h]: I think we should either call everything `TABLE` or everything `TBL`

File: qt/widgets/common/src/DataProcessorUI/QDataProcessorWidget.cpp [hunk 2]
```diff
@@ -306,31 +308,167 @@ void QDataProcessorWidget::selectAll() { ui.viewTable->selectAll(); }
...
+    DataProcessorAction toEnable) {
+  enableActionOnWidget(*m_contextMenu, toContextMenuIndex(toEnable));
+}
+
+int QDataProcessorWidget::toContextMenuIndex(DataProcessorAction action) {
+  return toCommandIndex(action);
+}
+
+int QDataProcessorWidget::toCommandIndex(DataProcessorAction action) {
+  switch (action) {
```
Comment (AntonPiccardoSelg) [inline qt/widgets/common/src/DataProcessorUI/QDataProcessorWidget.cpp]: Will this be an issue when you are dealing with non-grouped reduction scenarios?

**Review by AntonPiccardoSelg — CHANGES_REQUESTED** [review]
Nice refactoring work . Clear improvement of the code quality  :+1:.

There are only some minor point and a fairly important point regarding reduction scenarios which might not use post-processing.

**Comment by AntonPiccardoSelg** [comment]
@edwardb96 There seem to be merge conflicts on here

**Comment by edwardb96** [comment]
There's a lot more than merge conflicts wrong with this PR, #19951 its no longer scheduled for 3.11.
I'll be continuing work on this issue in a new branch `19951_prevent_external_table_modification_2`.

**Pull request closed without merging**